5_visualize/src/plot_metrics.py

```diff
@@ -0,0 +1,229 @@
+    :param train_predictions_filepath: Path to training set predictions csv (Default value = None)
+    :param include_train_mean: Whether to include metrics for the mean of the
+        training set in the plot (Default value = False).
+    :param doy_bin_width: Number of days to bin day of year to (Default value =
+        1). For example, a doy_bin_width of 7 means that the values in the
+        column doy_bin will increase by 7 at a time. Useful for smoothing out
+        plots of metrics by day of year.
+
+    """
+    # Dictionary relating inputs to column names
+    prediction_columns = {
+        'lake': 'site_id',
+        'depth': 'interpolated_depth',
+        'doy': 'doy_bin'
+    }
+    # Dictionary relating inputs to aggregating functions
+    agg_functions = {
+        'rmse': rms,
+        'bias': 'mean'
+    }
+
+    preds = load_predictions(predictions_filepath, doy_bin_width)
+
+    # Aggregate metric of residuals by a variable
+    agg_metric = (
+        preds.groupby(prediction_columns[plot_by])['residual']
+        .aggregate(agg_functions[metric])
+        .rename(f'LSTM_{metric}')
+    )
+
+    if plot_by == 'lake':
+        # Plotting by lake creates a geographic plot, so call separate function
+        p = plot_by_lake(agg_metric, lake_metadata_filepath, metric)
+    else:
+        if include_train_mean:
+            # For comparison, take the mean of the training set and use it as a predictor to get residuals
+            # This shows visually where prediction is more challenging due to observation variance
+            # or difference from the training data
+            train_preds = load_predictions(train_predictions_filepath, doy_bin_width)
+            train_mean_temp = train_preds.groupby(prediction_columns[plot_by])['temp'].mean().rename('train_mean_temp')
+            preds_train_mean = preds.merge(train_mean_temp, how='left', on=prediction_columns[plot_by]).copy()
+            preds_train_mean['train_mean_residual'] = preds_train_mean['temp'] - preds_train_mean['train_mean_temp']
+            metric_train_mean = (
+                preds_train_mean.groupby(prediction_columns[plot_by])['train_mean_residual']
+                .aggregate(agg_functions[metric])
+                .rename(f'train_mean_{metric}')
+            )
+            df_plot = pd.merge(agg_metric, metric_train_mean, how='outer', left_index=True, right_index=True)
+        else:
+            df_plot = agg_metric
+
+        zero_line = hv.HLine(0).opts(color='black', level='underlay')
+        if plot_by == 'depth':
+            if metric == 'bias':
+                p = df_plot.hvplot.line() * zero_line
+            else:
+                p = df_plot.hvplot.line()
+        elif plot_by == 'doy':
+            if metric == 'bias':
+                p = df_plot.hvplot.line() * zero_line
+            else:
+                p = df_plot.hvplot.line()
+        else:
+            raise ValueError(f'Evaluating metrics by {plot_by} is not supported')
```

Comment:
I'm confused about the conditional statements (lines 155 and 160) based on `plot_by` equaling `depth` or `doy`. It seems that they're identical. Would it be equivalent and more compact to having something like this instead:

```
if plot_by not in ['depth', 'doy']:
    raise ValueError(f'Evaluating metrics by {plot_by} is not supported')
else:
    if metric == 'bias':
        zero_line = hv.HLine(0).opts(color='black', level='underlay')
        p = df_plot.hvplot.line() * zero_line
    else:
        p = df_plot.hvplot.line()
```

If identical, the arrangement of lines is trivial, but, especially being unfamiliar with holoviews, the current arrangement makes me wonder if holoviews is accessing the `plot_by` object in some hidden/unclear way.

Reply:
Yes, you're right. That would be equivalent and more concise. There's no hidden use of `plot_by`. I think I'll leave it as it is for now, though, because I'll likely want the code arranged like this if/when I change to another plotting library. The plot axes will need to be labeled manually based on which variable we're plotting by.